stash/commits.go

```diff
@@ -0,0 +1,146 @@
+)
+
+var (
+	ErrorGetCommitMultipleItems = errors.New("multiple items returned for commit name")
+)
+
+const (
+	CommitsURI = "commits"
+)
+
+type Commit struct {
+	SessionInfo        `json:"sessionInfo,omitempty"`
+	Author             User   `json:"author,omitempty"`
+	AuthorTimestamp    int64  `json:"authorTimestamp,omitempty"`
+	Committer          User   `json:"committer,omitempty"`
+	CommitterTimestamp int64  `json:"committerTimestamp,omitempty"`
+	DisplayId          string `json:"displayId,omitempty"`
+	ID                 string `json:"id,omitempty"`
+	Message            string `json:"message,omitempty"`
+	Parents            []struct {
+		DisplayId string `json:"displayId,omitempty"`
+		ID        string `json:"id,omitempty"`
+	} `json:"parents,omitempty"`
+}
+
+func (r *Commit) Default() {
+}
+
+func (r *Commit) ValidateInfo() error {
+	return nil
+}
+
+func (r *Commit) Equals(actual gitprovider.InfoRequest) bool {
+	return reflect.DeepEqual(r, actual)
+}
+
+func NewStashCommits(client *stashClientImpl) StashCommits {
+	p := &Commits{
+		Paging:    Paging{},
+		Requester: client.Client(),
+		Commits:   make([]*Commit, 0),
+		log:       client.log,
+	}
+	return p
+}
+
+type StashCommits interface {
+	List(ctx context.Context, projectName, repoName string, opts *ListOptions) (*Paging, error)
+	Get(ctx context.Context, projectName, repoName, commitID string) (*Commit, error)
+	getCommits() []*Commit
+}
+
+type Commits struct {
+	StashCommits
+	Paging
+	httpclient.Requester
+	Commits []*Commit `json:"values,omitempty"`
+	log     logr.Logger
+}
+
+func (p *Commits) getCommits() []*Commit {
+	return p.Commits
+}
+
+func (p *Commits) List(ctx context.Context, projectName, repoName string, opts *ListOptions) (*Paging, error) {
+	var query *url.Values
+	query = addPaging(query, opts)
+	resp, err := p.Requester.Do(ctx, newURI(projectsURI, projectName, RepositoriesURI, repoName, CommitsURI), query, nil, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("list commits for repository failed, %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("list commits for repository api call returned unexpected status code: %d", resp.StatusCode)
+	}
+
+	respBody, err := httpclient.GetRespBody(resp)
+	if err != nil {
+		return nil, fmt.Errorf("list commits for repository failed, unable to obtain response body, %w", err)
+	}
+
+	if err := json.Unmarshal([]byte(respBody), &p); err != nil {
+		return nil, fmt.Errorf("list commits for repository failed, unable to unmarshall repository json, %w", err)
+	}
+
+	for _, commits := range p.getCommits() {
+		commits.setSessionInfo(resp)
+	}
+	return &p.Paging, nil
+}
+
+func (p *Commits) Get(ctx context.Context, projectName, repoName, commitID string) (*Commit, error) {
+	resp, err := p.Requester.Do(ctx, newURI(projectsURI, projectName, RepositoriesURI, repoName, CommitsURI, commitID), nil, nil, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("get commit failed, %w", err)
```

Comment:
I think this could include the information for the commit that the failure is referencing?